### PR TITLE
Added support to arbitrary project metadata key/value pairs

### PR DESCRIPTION
**Pull request by stooit**
# Checklist
- [X] Affected Issues have been mentioned in the Closing issues section
- [X] Documentation has been written/updated
- [X] PR title is ready for changelog and subsystem label(s) applied

At present categorising projects within Lagoon is unwieldy (env vars) and require external systems to collate and manage queries (e.g there is no way to directly add or query by arbitrary metadata).

This PR adds support for adding any number of key/value metadata pairs and querying where any project contains a specific key or key/value pair.

Sponsored by [GovCMS](https://www.govcms.gov.au/) and [Salsa Digital](https://www.salsadigital.com.au/)

From the docs added:
## Maintaining project metadata
Project metadata can be assigned using arbitrary key/value pairs. Projects can then be queried by the associated metadata; for example you may categorise projects by type of software, version number, or any other categorisation you may wish to query on later.
### Add/update metadata on a project
Updates to metadata expect a key/value pair. It operates as an `UPSERT`, meaning if a key already exists the value will be updated, otherwise inserted.
You may have any number of k/v pairs stored against a project.
```graphql
mutation {
  updateProjectMetadata(
    input: { id: 1,  patch: { key: "type", value: "saas" } }
  ) {
    id
    metadata
  }
}
```
### Query for projects by metadata
Queries may be by `key` only (e.g return all projects where a specific key exists) or both `key` and `value` where both key and value must match.
All projects that have the `version` tag:
```graphql
query projectsByMetadata {
  projectsByMetadata(metadata: [{key: "version"] ) {
    id
    name
  }
}
```
All projects that have the `version` tag, specifically version `8`:
```graphql
query projectsByMetadata {
  projectsByMetadata(metadata: [{key: "version", value: "8"] ) {
    id
    name
  }
}
```
### Removing metadata on a project
Metadata can be removed on a per-key basis. Other metadata key/value pairs will persist.
```graphql
mutation {
  removeProjectMetadataByKey (
    input: { id: 1,  key: "version" }
  ) {
    id
    metadata
  }
}
```

# Closing issues
Partial #1289, no UI

### CODE REVIEW

File: services/api/src/resources/project/resolvers.js [hunk 1]
```diff
@@ -193,6 +193,68 @@ const getProjectByName = async (
...
+) => {
+
+  let where;
+  try {
+    await hasPermission('project', 'viewAll');
+
+    where = whereAnd([
+      args.createdAfter ? 'created >= :created_after' : '',
+      args.gitUrl ? 'git_url = :git_url' : '',
+    ]);
```
Comment (Schnitzel) [inline services/api/src/resources/project/resolvers.js]: you don't accept `createdAfter` and `gitUrl` as parameters for the `projectsByMetadata` Query in GraphQL, so we don't need to check for them here.

File: services/api/src/resources/project/resolvers.js [hunk 2]
```diff
@@ -193,6 +193,68 @@ const getProjectByName = async (
...
+      return [];
+    }
+
+    const userProjectIds = await models.UserModel.getAllProjectsIdsForUser({
+      id: keycloakGrant.access_token.content.sub,
+    });
+
+    where = whereAnd([
+      args.createdAfter ? 'created >= :created_after' : '',
+      args.gitUrl ? 'git_url = :git_url' : '',
```
Comment (Schnitzel) [inline services/api/src/resources/project/resolvers.js]: not needed

File: services/api/src/resources/project/resolvers.js [hunk 3]
```diff
@@ -193,6 +193,68 @@ const getProjectByName = async (
...
+      inClause('id', userProjectIds),
+    ]);
+  }
+
+  for (const meta of args['metadata']) {
+
+    let key = meta['key'];
+    let val = meta['value'];
+
+    var q = `JSON_CONTAINS(metadata, '"`+val+`"', '$.`+key+`')`;
```
Comment (Schnitzel) [inline services/api/src/resources/project/resolvers.js:229]: wow, never heard of `JSON_CONTAINS` in MySQL? Crazy world :D

File: services/api/src/typeDefs.js [hunk 4]
```diff
@@ -340,6 +341,10 @@ const typeDefs = gql`
     Which groups are directly linked to project
     """
     groups: [GroupInterface]
+    """
+    Metadata key/values stored against a project
+    """
+    metadata: String
```
Comment (Schnitzel) [inline services/api/src/typeDefs.js]: there is a JSON field in GraphQL, see example:
https://github.com/amazeeio/lagoon/blob/8d95f023c12139b3672ce38582f5b0ebe9481cf2/services/api/src/typeDefs.js#L614
I think we can use that one?

File: services/api/src/util/db.js [hunk 5]
```diff
@@ -4,6 +4,8 @@ const camelcaseKeys = require('camelcase-keys');
 const R = require('ramda');
 const snakecase = require('./snakeCase');
 const snakecaseKeys = require('snakecase-keys');
+const logger = require('../logger');
```
Comment (Schnitzel) [inline services/api/src/util/db.js]: I don't think this is needed in here?

**Review by Schnitzel — CHANGES_REQUESTED** [review]

**Comment by Schnitzel** [comment]
wow, epic work @srowlands <3
one thing I'm not sure about the whole thing, we're adding a new GraphQL Query `getProjectsByMetadata`, but there is also the Query `allProjects` already that we could just extend that it supports to be filtered by Metadata.
I think this goes a bit into the architecture of GraphQL and might be a bit an esoteric question, but my gut feeling tells me that as a user of the GraphQL I would rater use an already existing query and just pass it more data to it.
Happy to hear the opinion of others though.

**Comment by seanhamlin** [comment]
❤️ this Stu! The only thing I would add is some simple graphQL tests to prove that the functionality works as designed.

e.g. 

* Add a key/value pair
* Query for it (assert can find 1 project)
* Remove it
* Query for it (assert cannot find any projects)

This can be picked up by the team as well, if you are slammed (let me know).

File: services/api/src/resources/project/resolvers.js (hunk 3 above)
Comment (stooit) [reply services/api/src/resources/project/resolvers.js:229]: ha yeah, sort-of relational/nosql hybrid

**Comment by stooit** [comment]
🙏 Thanks for the prompt review + feedback!

Have resolved the minor bits and pieces, keen to hear feedback on whether this should be rolled into the `allProjects` query too.

Agree on tests, I can circle back to this next mid next week unless someone picks that up in the meantime

**Review by Schnitzel — APPROVED** [review]

**Pull request merged**